### PR TITLE
Fix Erlang 24 Warnings

### PR DESCRIPTION
**Pull request by x1001100011**
There're 3 major points for this pr:
### http_uri -> uri_string
`http_uri:decode` -> `emqx_http_lib:uri_decode`
`http_uri:parse` -> `emqx_http_lib:uri_parse`
`http_uri:encode` -> `emqx_http_lib:uri_encode`

### pg2 -> pg
https://erlang.org/doc/man/pg.html
> Groups are automatically created when any process joins, and are removed when all processes leave the group. Non-existing group is considered empty (containing no processes).

### make test
`37 tests, 10 failures, 4 cancelled` for both master and this pr

### CODE REVIEW

File: src/coap_responder.erl [hunk 1]
```diff
@@ -233,12 +231,7 @@ handle_unobserve(_ChId, Request, Resource, State) ->
 cancel_observer(#coap_message{options=Options}, State=#state{module=Module, obstate=ObState}) ->
     ok = invoke_callback(Module, coap_unobserve, [ObState]),
     Uri = proplists:get_value(uri_path, Options, []),
-    ok = pg2:leave({coap_observer, Uri}, self()),
-    % will the last observer to leave this group please turn out the lights
```
Comment (zmstone) [inline src/coap_responder.erl:237]: does the new `pg` module does this for us ?

**Comment by zmstone** [comment]
https://github.com/emqx/gen_coap/blob/v0.3.2/src/gen_coap.app.src#L3
this is really confusing (git tag: 0.3.2, app version 0.2.1)
we should fix it while we are here.

File: src/coap_responder.erl (hunk 1 above)
Comment (x1001100011) [reply src/coap_responder.erl:237]: there's no `pg:delete/*` function, here's the docs from http://erlang.org/doc/man/pg.html
> Groups are automatically created when any process joins, and are removed when all processes leave the group. Non-existing group is considered empty (containing no processes).

and here's the code to delete the ets table once the last pid left
https://github.com/erlang/otp/blob/master/lib/kernel/src/pg.erl#L468

**Comment by x1001100011** [comment]
> https://github.com/emqx/gen_coap/blob/v0.3.2/src/gen_coap.app.src#L3
> this is really confusing (git tag: 0.3.2, app version 0.2.1)
> we should fix it while we are here.

could we focus on the erlang 24 warning on this pr? we can resolve other bugs, fix unit test, add the ci in separate prs.

File: rebar.config [hunk 2]
```diff
@@ -1,4 +1,6 @@
-{deps, [{esockd, {git, "https://github.com/emqx/esockd", {tag, "5.7.4"}}}]}.
+{deps, [ {esockd, {git, "https://github.com/emqx/esockd", {tag, "5.7.4"}}}
+       , {emqx_http_lib, {git, "https://github.com/emqx/emqx_http_lib", {tag, "0.1.0"}}}
```
Comment (x1001100011) [inline rebar.config]:
```suggestion
       , {emqx_http_lib, {git, "https://github.com/emqx/emqx_http_lib", {tag, "0.2.0"}}}
```

**Review by zmstone — APPROVED** [review]

File: src/client/coap_client.erl [hunk 3]
```diff
@@ -107,8 +107,15 @@ ack(Channel, Message) ->
...
-    {ok, {Scheme, _UserInfo, Host, PortNo, Path, Query}} =
-        http_uri:parse(Uri, [{scheme_defaults, [{coap, ?DEFAULT_COAP_PORT}, {coaps, ?DEFAULT_COAPS_PORT}]}]),
+    {ok, Parsed} = emqx_http_lib:uri_parse(Uri),
+    #{scheme := Scheme, host := Host, path := Path} = Parsed,
+    Query = maps:get(query, Parsed, ""),
+    DefaultPortNo =
+        case Scheme of
+            coap -> ?DEFAULT_COAP_PORT;
+            coaps -> ?DEFAULT_COAPS_PORT
+        end,
```
Comment (zmstone) [inline src/client/coap_client.erl]: default coap ports are added by `emqx_http_lib`
https://github.com/emqx/emqx_http_lib/blob/main/src/emqx_http_lib.erl#L132-L133

File: src/client/coap_client.erl [hunk 4]
```diff
@@ -117,7 +124,8 @@ split_path([$/]) -> [];
 split_path([$/ | Path]) -> split_segments(Path, $/, []).
 
 split_query([]) -> [];
-split_query([$? | Path]) -> split_segments(Path, $&, []).
+split_query([$? | Path]) -> split_segments(Path, $&, []);
```
Comment (zmstone) [inline src/client/coap_client.erl:122]: the new uri parser do not return query string with leading `?`, we can delete this clause.

**Review by zmstone — CHANGES_REQUESTED** [review]
added two coments

**Review by zmstone — APPROVED** [review]

**Comment by zmstone** [comment]
tagged v0.4.0

**Pull request merged**